File: 2024 June/Delete node in Doubly Linked List/main.py
# Link: https://www.geeksforgeeks.org/problems/delete-node-in-doubly-linked-list/1


class Node:

    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

    @staticmethod
    def print_list(node):
        while node is not None:
            print(node.data, end=" ")
            node = node.next
        print()


def createLinkedList(arr):
    head = Node(arr[0])
    last = head
    for value in arr[1:]:
        last.next = Node(value)
        last.next.prev = last
        last = last.next
    return head


class Solution:
    def delete_node(self, head, x):
        if x == 1:
            head = head.next
            head.prev = None
            return head

        cnt = 1
        temp = head
        while cnt < (x - 1):
            temp = temp.next
            cnt += 1
        temp.next = temp.next.next
        return head


if __name__ == "__main__":
    obj = Solution()
    head = createLinkedList([1, 3, 4])
    obj.delete_node(head, 3)
    Node.print_list(head)

    head = createLinkedList([1, 5, 2, 9])
    obj.delete_node(head, 1)
    Node.print_list(head)

File: 2024 June/Delete node in Doubly Linked List/main.cpp
// Link: https://www.geeksforgeeks.org/problems/delete-node-in-doubly-linked-list/1

#include <iostream>
#include <vector>
using namespace std;

struct Node
{
    int data;
    struct Node *next;
    struct Node *prev;

    Node(int x)
    {
        data = x;
        next = NULL;
        prev = NULL;
    }
};

void printList(struct Node *node)
{
    while (node != NULL)
    {
        printf("%d ", node->data);
        node = node->next;
    }
    printf("\n");
}

void createLinkedList(Node **head, vector<int> elements)
{
    Node *temp = *head;
    for (int i = 0; i < elements.size(); i++)
    {
        Node *newNode = new Node(elements[i]);
        if (*head == NULL)
        {
            *head = newNode;
            temp = newNode;
        }
        else
        {
            temp->next = newNode;
            newNode->prev = temp;
            temp = newNode;
        }
    }
}

class Solution
{
public:
    Node *deleteNode(Node *head, int x)
    {
        if (x == 1)
        {
            head = head->next;
            head->prev = NULL;
            return head;
        }
        int cnt = 1;
        Node *temp = head;
        while (cnt < x - 1)
        {
            temp = temp->next;
            cnt++;
        }
        temp->next = temp->next->next;
        return head;
    }
};

int main()
{
    Solution obj;
    Node *head = NULL;
    vector<int> elements = {1, 3, 4};
    createLinkedList(&head, elements);
    int x = 3;
    head = obj.deleteNode(head, x);
    printList(head);

    head = NULL;
    elements = {1, 5, 2, 9};
    createLinkedList(&head, elements);
    x = 1;
    head = obj.deleteNode(head, x);
    printList(head);
    return 0;
}